dvc/remote/s3.py

```diff
@@ -113,7 +116,33 @@ def copy(self, from_info, to_info, s3=None):
         s3 = s3 if s3 else self.s3
 
         source = {"Bucket": from_info["bucket"], "Key": from_info["path"]}
-        self.s3.copy(source, to_info["bucket"], to_info["path"])
+
+        head_object = self.get_head_object(
+            from_info["bucket"], from_info["path"]
+        )
+        etag = head_object["ETag"].strip('"')
+        if "-" in etag:  # Is Multipart
+            head_object_first_part = self.get_head_object(
+                source["Bucket"], source["Key"], PartNumber=1
+            )
+            required_chunksize = head_object_first_part["ContentLength"]
+            transfer_config = boto3.s3.transfer.TransferConfig(
+                multipart_chunksize=required_chunksize
+            )
+        else:
+            transfer_config = boto3.s3.transfer.TransferConfig(
+                multipart_threshold=head_object["ContentLength"] + 1
+            )  # Default
+
+        s3.copy(
+            source, to_info["bucket"], to_info["path"], Config=transfer_config
+        )
+
+        cached_etag = self.get_etag(to_info["bucket"], to_info["path"])
+        if etag != cached_etag:
+            msg = "ETag mismatch detected when copying file to cache!\
+                         (expected: '{}', actual: '{}')"
+            raise DvcException(msg.format(etag, cached_etag))
```

Comment:
This could be a separate exception extending DvcException in dvc/exceptions.py, for example ETagMismatchException that would accept etag and cached_etag as init arguments. Would you mind changing that?

Reply:
Done